File: engine.ts
export type CandleInterval = '1h' | '4h' | '1d';
export type ZoneType = 'order_block' | 'volume_zone' | 'sideways_box';
export type StrategyStatus = 'active' | 'reacted' | 'invalidated' | 'canceled';

export interface Candle {
    openTime: number; open: number; high: number; low: number; close: number; volume: number; closeTime: number; isBullish: boolean;
}

export interface ScoreBreakdown {
    total: number; structure?: number; continuation?: number; breakoutPrep?: number; turningPoint?: number;
    insideBarBreakout?: number; engulfingBreakout?: number; pinbarReversal?: number; rsiExtreme?: number; regularDivergence?: number; volumeBonus?: number;
}

export interface Zone {
    id: string; type: ZoneType; startIndex: number; endIndex: number; direction: 'long' | 'short'; score: ScoreBreakdown; status: StrategyStatus; createdAt: number; createdIndex: number;
}

export interface Box extends Zone {
    type: 'sideways_box'; archetype: 'continuation_box' | 'breakout_prep_box' | 'turning_point_base' | 'unknown';
    breakoutIndex: number; high: number; low: number; ep: number; sl: number; tp: number; touchedAt?: number; resolvedAt?: number; realizedPnl?: number; realizedPnlPercent?: number; isEntered?: boolean; enteredAt?: number; assetRoiPercent?: number; positionSize?: number; riskAmount?: number; marginUsed?: number; skipReason?: 'max_positions' | 'no_margin' | 'sl_before_ep';
}

export class BinanceAPI {
    private static readonly BASE_URL = 'https://api.binance.com/api/v3';
    static async fetchKlines(symbol: string, interval: CandleInterval, startTime?: number, endTime?: number, limit = 1000): Promise<{data: Candle[], isMock: boolean}> {
        let allCandles: Candle[] = [];
        let currentStartTime = startTime;
        let fetchCount = 0;
        const MAX_FETCHES = 30; 
        try {
            while (fetchCount < MAX_FETCHES) {
                let url = `${this.BASE_URL}/klines?symbol=${symbol}&interval=${interval}&limit=${limit}`;
                if (currentStartTime) url += `&startTime=${currentStartTime}`;
                if (endTime) url += `&endTime=${endTime}`;
                const response = await fetch(url);
                if (!response.ok) throw new Error(`Binance API error: ${response.statusText}`);
                const rawData: any[][] = await response.json();
                if (rawData.length === 0) break; 
                const data = rawData.map(row => {
                    const open = parseFloat(row[1]); const close = parseFloat(row[4]);
                    return { openTime: row[0], open, high: parseFloat(row[2]), low: parseFloat(row[3]), close, volume: parseFloat(row[5]), closeTime: row[6], isBullish: close >= open };
                });
                allCandles = allCandles.concat(data);
                const lastCandle = data[data.length - 1];
                if (endTime && lastCandle.closeTime >= endTime) break;
                if (rawData.length < limit) break;
                currentStartTime = lastCandle.closeTime + 1;
                fetchCount++;
                await new Promise(resolve => setTimeout(resolve, 50));
            }
            return { data: allCandles, isMock: false };
        } catch (error) {
            console.warn('API 호출 실패, 빈 배열을 반환합니다.', error);
            return { data: [], isMock: true };
        }
    }
}

export class Indicators {
    static calculateRSI(candles: Candle[], period = 14): number[] {
        const rsi = new Array(candles.length).fill(0);
        if (candles.length <= period) return rsi;
        let avgGain = 0; let avgLoss = 0;
        for (let i = 1; i <= period; i++) {
            const change = candles[i].close - candles[i - 1].close;
            if (change > 0) avgGain += change; else avgLoss -= change;
        }
        avgGain /= period; avgLoss /= period;
        rsi[period] = 100 - (100 / (1 + (avgGain / (avgLoss || 1e-10))));
        for (let i = period + 1; i < candles.length; i++) {
            const change = candles[i].close - candles[i - 1].close;
            const gain = change > 0 ? change : 0; const loss = change < 0 ? -change : 0;
            avgGain = ((avgGain * (period - 1)) + gain) / period; avgLoss = ((avgLoss * (period - 1)) + loss) / period;
            rsi[i] = 100 - (100 / (1 + (avgGain / (avgLoss || 1e-10))));
        }
        return rsi;
    }
    static hasRegularDivergence(candles: Candle[], rsi: number[], index: number, isBullishBreakout: boolean): boolean {
        if (index < 10) return false;
        const isPivotLow = (idx: number) => {
            if (idx < 3 || idx > candles.length - 4) return false;
            const l = candles[idx].low; return l <= candles[idx-1].low && l <= candles[idx-2].low && l <= candles[idx-3].low && l <= candles[idx+1].low && l <= candles[idx+2].low && l <= candles[idx+3].low;
        };
        const isPivotHigh = (idx: number) => {
            if (idx < 3 || idx > candles.length - 4) return false;
            const h = candles[idx].high; return h >= candles[idx-1].high && h >= candles[idx-2].high && h >= candles[idx-3].high && h >= candles[idx+1].high && h >= candles[idx+2].high && h >= candles[idx+3].high;
        };
        const pivots: number[] = [];
        for (let i = index; i >= index - 30 && pivots.length < 2; i--) {
            if (isBullishBreakout ? isPivotLow(i) : isPivotHigh(i)) pivots.push(i);
        }
        if (pivots.length < 2) return false;
        const [recent, previous] = pivots;
        if (isBullishBreakout) return candles[recent].low < candles[previous].low && rsi[recent] > rsi[previous];
        else return candles[recent].high > candles[previous].high && rsi[recent] < rsi[previous];
    }
    static hasVolumeExpansion(candles: Candle[], currentIndex: number): boolean {
        if (currentIndex < 35) return false;
        let recentVol = 0; let pastVol = 0;
        for (let i = currentIndex - 4; i <= currentIndex; i++) recentVol += candles[i].volume;
        for (let i = currentIndex - 34; i <= currentIndex - 5; i++) pastVol += candles[i].volume;
        return recentVol > pastVol;
    }
    static isPinbar(candle: Candle): boolean {
        const body = Math.abs(candle.open - candle.close);
        const upperWick = candle.high - Math.max(candle.open, candle.close);
        const lowerWick = Math.min(candle.open, candle.close) - candle.low;
        return (lowerWick >= 2 * body && upperWick <= 0.5 * body) || (upperWick >= 2 * body && lowerWick <= 0.5 * body);
    }
    static isEngulfing(prev: Candle, curr: Candle): boolean {
        if (prev.isBullish === curr.isBullish) return false;
        const prevBody = Math.abs(prev.open - prev.close); const currBody = Math.abs(curr.open - curr.close);
        return currBody > prevBody && Math.max(curr.open, curr.close) >= Math.max(prev.open, prev.close) && Math.min(curr.open, curr.close) <= Math.min(prev.open, prev.close);
    }
    static isInsideBar(prev: Candle, curr: Candle): boolean {
        return curr.high <= prev.high && curr.low >= prev.low;
    }
}

export class SidewaysBoxDetector {
    private readonly PASS_THRESHOLD = 70;
    public detect(candles: Candle[], interval: CandleInterval, rrRatio: number): Box[] {
        const boxes: Box[] = [];
        const rsi = Indicators.calculateRSI(candles);
        for (let i = 35; i < candles.length - 4; i++) {
            let high = candles[i].high; let low = candles[i].low;
            let bIndex = -1; let direction: 'long' | 'short' | null = null;
            for (let j = i + 1; j < candles.length - 3; j++) {
                if (candles[j].close > high) { bIndex = j; direction = 'long'; break; }
                if (candles[j].close < low) { bIndex = j; direction = 'short'; break; }
                high = Math.max(high, candles[j].high); low = Math.min(low, candles[j].low);
            }
            if (bIndex === -1 || !direction) continue;
            if (bIndex - i < 1) continue;
            if (!this.validatePreBreakReentry(candles, i, bIndex, direction)) continue;
            const archetype = this.determineArchetype(candles, i, bIndex, bIndex - i, interval, direction);
            if (archetype === 'unknown') continue;
            const score = this.calculateScore(candles, rsi, bIndex, archetype, direction);
            if (score.total < this.PASS_THRESHOLD) continue;
            const ep = (high + low) / 2;
            let sl: number, tp: number;
            if (direction === 'long') { sl = low; tp = ep + rrRatio * (ep - sl); } else { sl = high; tp = ep - rrRatio * (sl - ep); }
            boxes.push({ id: `box_${candles[bIndex].openTime}`, type: 'sideways_box', archetype, direction, startIndex: i, endIndex: bIndex - 1, breakoutIndex: bIndex, createdIndex: bIndex + 3, createdAt: candles[bIndex + 3].openTime, high, low, ep, sl, tp, score, status: 'active' });
            i = bIndex; 
        }
        return boxes;
    }
    private validatePreBreakReentry(candles: Candle[], startIdx: number, bIndex: number, dir: 'long'|'short'): boolean {
        let obCandle: Candle | null = null;
        for (let k = bIndex - 1; k >= startIdx; k--) { if (candles[k].isBullish !== (dir === 'long')) { obCandle = candles[k]; break; } }
        if (!obCandle) return false;
        for (let k = bIndex + 1; k <= bIndex + 3; k++) {
            if (k >= candles.length) return false;
            if (candles[k].low <= Math.max(obCandle.open, obCandle.close) && candles[k].high >= Math.min(obCandle.open, obCandle.close)) return false;
        }
        return true;
    }
    private determineArchetype(candles: Candle[], startIdx: number, bIdx: number, len: number, interval: CandleInterval, dir: 'long'|'short'): Box['archetype'] {
        const c1 = candles[bIdx - 2]; const c2 = candles[bIdx - 1];
        if (len >= 10 && Indicators.hasVolumeExpansion(candles, bIdx - 1)) {
            let hasReversalSign = false;
            for(let k = startIdx; k < bIdx; k++) { if (Indicators.isPinbar(candles[k]) || (k > startIdx && Indicators.isEngulfing(candles[k-1], candles[k]))) { hasReversalSign = true; break; } }
            if (hasReversalSign) return 'turning_point_base';
        }
        if (len >= 10 && c1 && c2 && c1.isBullish !== c2.isBullish && Indicators.isEngulfing(c1, c2)) return 'breakout_prep_box';
        if ((interval === '1h' && len >= 10 && len <= 40) || (interval === '4h' && len >= 5 && len <= 15) || (interval === '1d' && len >= 1 && len <= 3)) return 'continuation_box';
        return 'unknown';
    }
    private calculateScore(candles: Candle[], rsi: number[], bIdx: number, arch: string, dir: 'long'|'short'): ScoreBreakdown {
        const s: ScoreBreakdown = { total: 60, structure: 60 };
        if (arch === 'continuation_box') { s.continuation = 10; s.total += 10; }
        if (arch === 'breakout_prep_box') { s.breakoutPrep = 15; s.total += 15; }
        if (arch === 'turning_point_base') { s.turningPoint = 20; s.total += 20; }
        if (Indicators.isInsideBar(candles[bIdx - 2], candles[bIdx - 1])) { s.insideBarBreakout = 5; s.total += 5; }
        if (Indicators.isEngulfing(candles[bIdx - 1], candles[bIdx])) { s.engulfingBreakout = 8; s.total += 8; }
        if (Indicators.isPinbar(candles[bIdx])) { s.pinbarReversal = 8; s.total += 8; }
        if (rsi[bIdx] <= 30 || rsi[bIdx] >= 70) { s.rsiExtreme = 8; s.total += 8; }
        if (Indicators.hasRegularDivergence(candles, rsi, bIdx, dir === 'long')) { s.regularDivergence = 12; s.total += 12; }
        if (Indicators.hasVolumeExpansion(candles, bIdx)) { s.volumeBonus = 10; s.total += 10; }
        return s;
    }
}

export class ChartEngine {
    private detector = new SidewaysBoxDetector();
    public process(candles: Candle[], interval: CandleInterval, rrRatio: number): Box[] {
        return this.dedupe(this.detector.detect(candles, interval, rrRatio));
    }
    private dedupe(boxes: Box[]): Box[] {
        const filtered: Box[] = [];
        for (const box of boxes) {
            const overlap = filtered.find(f => f.direction === box.direction && ((box.startIndex >= f.startIndex && box.startIndex <= f.endIndex) || (box.endIndex >= f.startIndex && box.endIndex <= f.endIndex)));
            if (!overlap) filtered.push(box);
            else if (box.startIndex < overlap.startIndex || (box.startIndex === overlap.startIndex && box.score.total > overlap.score.total)) { filtered[filtered.indexOf(overlap)] = box; }
        }
        return filtered;
    }
}

File: bot_py/engine.py
"""코어 차트 분석 엔진 (bot/engine.ts 이식)."""
from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Literal, Optional

import requests

CandleInterval = Literal["1h", "4h", "1d"]
ZoneType = Literal["order_block", "volume_zone", "sideways_box"]
StrategyStatus = Literal["active", "reacted", "invalidated", "canceled"]
Archetype = Literal["continuation_box", "breakout_prep_box", "turning_point_base", "unknown"]
Direction = Literal["long", "short"]


@dataclass
class Candle:
    open_time: int
    open: float
    high: float
    low: float
    close: float
    volume: float
    close_time: int
    is_bullish: bool


@dataclass
class ScoreBreakdown:
    total: int = 0
    structure: Optional[int] = None
    continuation: Optional[int] = None
    breakout_prep: Optional[int] = None
    turning_point: Optional[int] = None
    inside_bar_breakout: Optional[int] = None
    engulfing_breakout: Optional[int] = None
    pinbar_reversal: Optional[int] = None
    rsi_extreme: Optional[int] = None
    regular_divergence: Optional[int] = None
    volume_bonus: Optional[int] = None


@dataclass
class Box:
    id: str
    type: ZoneType
    archetype: Archetype
    start_index: int
    end_index: int
    direction: Direction
    score: ScoreBreakdown
    status: StrategyStatus
    created_at: int
    created_index: int
    breakout_index: int
    high: float
    low: float
    ep: float
    sl: float
    tp: float
    touched_at: Optional[int] = None
    resolved_at: Optional[int] = None
    realized_pnl: Optional[float] = None
    realized_pnl_percent: Optional[float] = None
    is_entered: bool = False
    entered_at: Optional[int] = None
    asset_roi_percent: Optional[float] = None
    position_size: Optional[float] = None
    risk_amount: Optional[float] = None
    margin_used: Optional[float] = None
    skip_reason: Optional[Literal["max_positions", "no_margin", "sl_before_ep"]] = None
    quantity: Optional[float] = None
    entered_price: Optional[float] = None


class BinanceAPI:
    BASE_URL = "https://api.binance.com/api/v3"

    @classmethod
    def fetch_klines(
        cls,
        symbol: str,
        interval: str,
        start_time: Optional[int] = None,
        end_time: Optional[int] = None,
        limit: int = 1000,
    ) -> tuple[list[Candle], bool]:
        all_candles: list[Candle] = []
        current_start_time = start_time
        fetch_count = 0
        max_fetches = 30
        try:
            while fetch_count < max_fetches:
                params: dict[str, object] = {"symbol": symbol, "interval": interval, "limit": limit}
                if current_start_time:
                    params["startTime"] = current_start_time
                if end_time:
                    params["endTime"] = end_time
                response = requests.get(f"{cls.BASE_URL}/klines", params=params, timeout=15)
                if not response.ok:
                    raise RuntimeError(f"Binance API error: {response.status_code} {response.reason}")
                raw_data = response.json()
                if len(raw_data) == 0:
                    break
                data = [
                    Candle(
                        open_time=row[0],
                        open=float(row[1]),
                        high=float(row[2]),
                        low=float(row[3]),
                        close=float(row[4]),
                        volume=float(row[5]),
                        close_time=row[6],
                        is_bullish=float(row[4]) >= float(row[1]),
                    )
                    for row in raw_data
                ]
                all_candles.extend(data)
                last_candle = data[-1]
                if end_time and last_candle.close_time >= end_time:
                    break
                if len(raw_data) < limit:
                    break
                current_start_time = last_candle.close_time + 1
                fetch_count += 1
                time.sleep(0.05)
            return all_candles, False
        except Exception as error:
            print(f"API 호출 실패, 빈 배열을 반환합니다. {error}")
            return [], True


class Indicators:
    @staticmethod
    def calculate_rsi(candles: list[Candle], period: int = 14) -> list[float]:
        rsi = [0.0] * len(candles)
        if len(candles) <= period:
            return rsi
        avg_gain = 0.0
        avg_loss = 0.0
        for i in range(1, period + 1):
            change = candles[i].close - candles[i - 1].close
            if change > 0:
                avg_gain += change
            else:
                avg_loss -= change
        avg_gain /= period
        avg_loss /= period
        rsi[period] = 100 - (100 / (1 + (avg_gain / (avg_loss or 1e-10))))
        for i in range(period + 1, len(candles)):
            change = candles[i].close - candles[i - 1].close
            gain = change if change > 0 else 0
            loss = -change if change < 0 else 0
            avg_gain = ((avg_gain * (period - 1)) + gain) / period
            avg_loss = ((avg_loss * (period - 1)) + loss) / period
            rsi[i] = 100 - (100 / (1 + (avg_gain / (avg_loss or 1e-10))))
        return rsi

    @staticmethod
    def has_regular_divergence(
        candles: list[Candle], rsi: list[float], index: int, is_bullish_breakout: bool
    ) -> bool:
        if index < 10:
            return False

        def is_pivot_low(idx: int) -> bool:
            if idx < 3 or idx > len(candles) - 4:
                return False
            low = candles[idx].low
            return (
                low <= candles[idx - 1].low
                and low <= candles[idx - 2].low
                and low <= candles[idx - 3].low
                and low <= candles[idx + 1].low
                and low <= candles[idx + 2].low
                and low <= candles[idx + 3].low
            )

        def is_pivot_high(idx: int) -> bool:
            if idx < 3 or idx > len(candles) - 4:
                return False
            high = candles[idx].high
            return (
                high >= candles[idx - 1].high
                and high >= candles[idx - 2].high
                and high >= candles[idx - 3].high
                and high >= candles[idx + 1].high
                and high >= candles[idx + 2].high
                and high >= candles[idx + 3].high
            )

        pivots: list[int] = []
        i = index
        while i >= index - 30 and len(pivots) < 2:
            if is_pivot_low(i) if is_bullish_breakout else is_pivot_high(i):
                pivots.append(i)
            i -= 1

        if len(pivots) < 2:
            return False

        recent, previous = pivots
        if is_bullish_breakout:
            return candles[recent].low < candles[previous].low and rsi[recent] > rsi[previous]
        return candles[recent].high > candles[previous].high and rsi[recent] < rsi[previous]

    @staticmethod
    def has_volume_expansion(candles: list[Candle], current_index: int) -> bool:
        if current_index < 35:
            return False
        recent_vol = sum(c.volume for c in candles[current_index - 4 : current_index + 1])
        past_vol = sum(c.volume for c in candles[current_index - 34 : current_index - 4])
        return recent_vol > past_vol

    @staticmethod
    def is_pinbar(candle: Candle) -> bool:
        body = abs(candle.open - candle.close)
        upper_wick = candle.high - max(candle.open, candle.close)
        lower_wick = min(candle.open, candle.close) - candle.low
        return (lower_wick >= 2 * body and upper_wick <= 0.5 * body) or (
            upper_wick >= 2 * body and lower_wick <= 0.5 * body
        )

    @staticmethod
    def is_engulfing(prev: Candle, curr: Candle) -> bool:
        if prev.is_bullish == curr.is_bullish:
            return False
        prev_body = abs(prev.open - prev.close)
        curr_body = abs(curr.open - curr.close)
        return (
            curr_body > prev_body
            and max(curr.open, curr.close) >= max(prev.open, prev.close)
            and min(curr.open, curr.close) <= min(prev.open, prev.close)
        )

    @staticmethod
    def is_inside_bar(prev: Candle, curr: Candle) -> bool:
        return curr.high <= prev.high and curr.low >= prev.low


class SidewaysBoxDetector:
    PASS_THRESHOLD = 70

    def detect(self, candles: list[Candle], interval: str, rr_ratio: float) -> list[Box]:
        boxes: list[Box] = []
        rsi = Indicators.calculate_rsi(candles)
        i = 35
        while i < len(candles) - 4:
            high = candles[i].high
            low = candles[i].low
            b_index = -1
            direction: Optional[Direction] = None
            for j in range(i + 1, len(candles) - 3):
                if candles[j].close > high:
                    b_index, direction = j, "long"
                    break
                if candles[j].close < low:
                    b_index, direction = j, "short"
                    break
                high = max(high, candles[j].high)
                low = min(low, candles[j].low)

            if b_index == -1 or direction is None:
                i += 1
                continue
            if b_index - i < 1:
                i += 1
                continue
            if not self._validate_pre_break_reentry(candles, i, b_index, direction):
                i += 1
                continue
            archetype = self._determine_archetype(candles, i, b_index, b_index - i, interval, direction)
            if archetype == "unknown":
                i += 1
                continue
            score = self._calculate_score(candles, rsi, b_index, archetype, direction)
            if score.total < self.PASS_THRESHOLD:
                i += 1
                continue

            ep = (high + low) / 2
            if direction == "long":
                sl = low
                tp = ep + rr_ratio * (ep - sl)
            else:
                sl = high
                tp = ep - rr_ratio * (sl - ep)

            boxes.append(
                Box(
                    id=f"box_{candles[b_index].open_time}",
                    type="sideways_box",
                    archetype=archetype,
                    direction=direction,
                    start_index=i,
                    end_index=b_index - 1,
                    breakout_index=b_index,
                    created_index=b_index + 3,
                    created_at=candles[b_index + 3].open_time,
                    high=high,
                    low=low,
                    ep=ep,
                    sl=sl,
                    tp=tp,
                    score=score,
                    status="active",
                )
            )
            i = b_index + 1
        return boxes

    @staticmethod
    def _validate_pre_break_reentry(
        candles: list[Candle], start_idx: int, b_index: int, direction: Direction
    ) -> bool:
        ob_candle: Optional[Candle] = None
        for k in range(b_index - 1, start_idx - 1, -1):
            if candles[k].is_bullish != (direction == "long"):
                ob_candle = candles[k]
                break
        if ob_candle is None:
            return False
        for k in range(b_index + 1, b_index + 4):
            if k >= len(candles):
                return False
            if candles[k].low <= max(ob_candle.open, ob_candle.close) and candles[k].high >= min(
                ob_candle.open, ob_candle.close
            ):
                return False
        return True

    @staticmethod
    def _determine_archetype(
        candles: list[Candle], start_idx: int, b_idx: int, length: int, interval: str, direction: Direction
    ) -> Archetype:
        c1 = candles[b_idx - 2]
        c2 = candles[b_idx - 1]
        if length >= 10 and Indicators.has_volume_expansion(candles, b_idx - 1):
            has_reversal_sign = False
            for k in range(start_idx, b_idx):
                if Indicators.is_pinbar(candles[k]) or (
                    k > start_idx and Indicators.is_engulfing(candles[k - 1], candles[k])
                ):
                    has_reversal_sign = True
                    break
            if has_reversal_sign:
                return "turning_point_base"
        if length >= 10 and c1.is_bullish != c2.is_bullish and Indicators.is_engulfing(c1, c2):
            return "breakout_prep_box"
        if (
            (interval == "1h" and 10 <= length <= 40)
            or (interval == "4h" and 5 <= length <= 15)
            or (interval == "1d" and 1 <= length <= 3)
        ):
            return "continuation_box"
        return "unknown"

    @staticmethod
    def _calculate_score(
        candles: list[Candle], rsi: list[float], b_idx: int, arch: Archetype, direction: Direction
    ) -> ScoreBreakdown:
        s = ScoreBreakdown(total=60, structure=60)
        if arch == "continuation_box":
            s.continuation = 10
            s.total += 10
        if arch == "breakout_prep_box":
            s.breakout_prep = 15
            s.total += 15
        if arch == "turning_point_base":
            s.turning_point = 20
            s.total += 20
        if Indicators.is_inside_bar(candles[b_idx - 2], candles[b_idx - 1]):
            s.inside_bar_breakout = 5
            s.total += 5
        if Indicators.is_engulfing(candles[b_idx - 1], candles[b_idx]):
            s.engulfing_breakout = 8
            s.total += 8
        if Indicators.is_pinbar(candles[b_idx]):
            s.pinbar_reversal = 8
            s.total += 8
        if rsi[b_idx] <= 30 or rsi[b_idx] >= 70:
            s.rsi_extreme = 8
            s.total += 8
        if Indicators.has_regular_divergence(candles, rsi, b_idx, direction == "long"):
            s.regular_divergence = 12
            s.total += 12
        if Indicators.has_volume_expansion(candles, b_idx):
            s.volume_bonus = 10
            s.total += 10
        return s


class ChartEngine:
    def __init__(self) -> None:
        self._detector = SidewaysBoxDetector()

    def process(self, candles: list[Candle], interval: str, rr_ratio: float) -> list[Box]:
        return self._dedupe(self._detector.detect(candles, interval, rr_ratio))

    @staticmethod
    def _dedupe(boxes: list[Box]) -> list[Box]:
        filtered: list[Box] = []
        for box in boxes:
            overlap = next(
                (
                    f
                    for f in filtered
                    if f.direction == box.direction
                    and (
                        (box.start_index >= f.start_index and box.start_index <= f.end_index)
                        or (box.end_index >= f.start_index and box.end_index <= f.end_index)
                    )
                ),
                None,
            )
            if overlap is None:
                filtered.append(box)
            elif box.start_index < overlap.start_index or (
                box.start_index == overlap.start_index and box.score.total > overlap.score.total
            ):
                filtered[filtered.index(overlap)] = box
        return filtered
